scipy/spatial/ckdtree.pyi

```diff
@@ -1,9 +1,221 @@
+    @property
+    def data(self) -> np.ndarray: ...
+    @property
+    def maxes(self) -> np.ndarray: ...
+    @property
+    def mins(self) -> np.ndarray: ...
+    @property
+    def indices(self) -> np.ndarray: ...
+    @property
+    def boxsize(self) -> _BoxType: ...
+
+    # NOTE: In practice `__init__` is used as constructor, not `__new__`.
+    # The latter gives us more flexibility in setting the generic parameter
+    # though.
+    @overload
+    def __new__(  # type: ignore[misc]
+        cls,
+        data: npt.ArrayLike,
+        leafsize: int = ...,
+        compact_nodes: bool = ...,
+        copy_data: bool = ...,
+        balanced_tree: bool = ...,
+        boxsize: None = ...,
+    ) -> cKDTree[None]: ...
+    @overload
+    def __new__(
+        cls,
+        data: npt.ArrayLike,
+        leafsize: int = ...,
+        compact_nodes: bool = ...,
+        copy_data: bool = ...,
+        balanced_tree: bool = ...,
+        boxsize: npt.ArrayLike = ...,
+    ) -> cKDTree[np.ndarray]: ...
+
+    # TODO: returns a 2-tuple of scalars if `x.ndim == 1` and `k == 1`,
+    # returns a 2-tuple of arrays otherwise
+    def query(
+        self,
+        x: npt.ArrayLike,
+        k: npt.ArrayLike = ...,
+        eps: float = ...,
+        p: float = ...,
+        distance_upper_bound: float = ...,
+        workers: Optional[int] = ...,
+    ) -> Tuple[Any, Any]: ...
+
+    # TODO: returns a list scalars if `x.ndim <= 1`,
+    # returns an object array of lists otherwise
+    def query_ball_point(
+        self,
+        x: npt.ArrayLike,
+        r: npt.ArrayLike,
+        p: float,
+        eps: float = ...,
+        workers: Optional[int] = ...,
+        return_sorted: Optional[bool] = ...,
+        return_length: bool = ...
+    ) -> Any: ...
+
+    def query_ball_tree(
+        self,
+        other: cKDTree,
+        r: float,
+        p: float,
+        eps: float = ...,
+    ) -> List[List[int]]: ...
+
+    @overload
+    def query_pairs(  # type: ignore[misc]
+        self,
+        r: float,
+        p: float = ...,
+        eps: float = ...,
+        output_type: Literal["set"] = ...,
+    ) -> Set[Tuple[int, int]]: ...
+    @overload
+    def query_pairs(
+        self,
+        r: float,
+        p: float = ...,
+        eps: float = ...,
+        output_type: Literal["ndarray"] = ...,
+    ) -> np.ndarray: ...
+
+    @overload
+    def count_neighbors(  # type: ignore[misc]
+        self,
+        other: cKDTree,
+        r: _ArrayLike0D,
+        p: float = ...,
+        weights: None = ...,
+        cumulative: bool = ...,
+    ) -> int: ...
+    @overload
+    def count_neighbors(  # type: ignore[misc]
+        self,
+        other: cKDTree,
+        r: _ArrayLike0D,
+        p: float = ...,
+        weights: _WeightType = ...,
+        cumulative: bool = ...,
+    ) -> np.float64: ...
+    @overload
+    def count_neighbors(
+        self,
+        other: cKDTree,
+        r: npt.ArrayLike,
```

Comment:
I see that the `type: ignore` comments are silencing errors about overload ambiguity. Is there an ambiguity between `ArrayLike` and `_ArrayLike0D` here, or is it something else? 

Whatever the cause, I think that if the typing is ambiguous then doesn't returning `Union` make more sense? I think it's better for the typing to be correct but general, than to be specific but wrong.

Reply:
In this specific case there is no ambiguity, everything is by design here. 
`_ArrayLike0D` is still a subtype of `ArrayLike` though, so there is annoyingly enough no way to get rid of the error except by silencing it,

Reply:
Mypy (and type checkers in general) handle a "first-come-first-served" principle with regards to overload matching,
hence why patterns why "subtype-followed-by-supertype" work just fine in practice.

> And if 0d numpy arrays give the wrong typing results, isn't that an issue?

Yes, and it is an annoyingly painful issue, one that is prevalent for 0D arrays in general.
Currently there are three possible ways of dealing with automatic 0D->Scalar casts (I'd hardly call them solutions though...):

* Return `Any`: This removes all type safety, effectivelly introducing a big false negative.
* Return an unsafe union: The opposite issue of `Any`; introduces false positives.
* Ignore automatic  0D -> scalar casts: Introduces false positives, but only for 0D arrays. This IMO the lesser of evils, as 0D arrays are used comparatively little compared to normal scalars and >= 1D arrays.